### ISSUE TITLE
3

### ISSUE THREAD

**Issue by HeSicheng00** [HeSicheng00/Mason-HE#3]
## **Summary:** 

After I improved the image-to-image DDPM, its generation and classification results have been greatly improved. Among them, the classification result is much better than unconditional DDPM, and it also exceeds the discriminative models including pyramidnet, resnet, and so on.

### A. Improvements over previous model:
To solve the previous problem, this time the model combines the features of the conditional image with the input noise in two parts:
1. before the noisy image is fed into the DDPM, it is fused with the features of the conditional image by a convolutional neural network;
2. after the noisy image is input to DDPM, the learning of the conditioned image features is performed by cross attention.

### B. Diffusion model architectures:

**I tried the following three architectures and found that the last one worked best:**
- [ ] **Structure 1**: Concat condition image and noise, use self attention in the UNet blocks.
- [ ] **Structure 2**: In addition to the above structure, cross attention is added in middle channel of UNet.
- [x] **Structure 3**: Cross attention is used in more UNet blocks and condtion image is fed into the feature extraction backbone.

The detailed structure of UNet is shown below (only the Encoder is drawn):
![b8b6b0e298079734e1fd9bc75a0ac2f](https://github.com/kyotovision/he_sicheng/assets/56251275/0af0cfbc-3272-4e1b-a5a8-4f3a3dd63ee6)

### C. Results:

After 10 epochs:
![4ddfa527e99c3367d75f7e91a891ba9](https://github.com/kyotovision/he_sicheng/assets/56251275/bec11e2a-5473-454a-8178-b9c0746239dc)

### A. Classification Network Details:

Following the previous paper(e.g., [Diffusion Models Beat GANs on Image Classification](https://arxiv.org/pdf/2307.08702v1.pdf)), we mainly use the diffusion model as a feature extraction network on which only the classification head is trained. 

However, a diffusion model corresponds to **different time steps** as well as **different UNet blocks**. We need to choose the appropriate time step and the UNet block. Also, a right **classification head** is needed as well.

1.  **different time steps**: 
When training the diffusion model, different time steps correspond to different noise levels imposed on x0 (the original image).
However, since the classification process does not involve the ground truth (standard font) in the diffusion model. So at this point the default time step is located at the maximum value, i.e. the input to the diffusion model is a fully Gaussian noise.

2.  **different UNet blocks**: 
The output of the bottleneck as well as each UNet block in the decoder can be used as the extracted features.
![4b5519c7bfc42b79d6bd308801e796d](https://github.com/kyotovision/he_sicheng/assets/56251275/fc28d128-3916-4514-8f57-9dea49a9cbdb)
According to experiments, the accuracy of features generated by block numbered between 3 and 7 is higher.

4.  **classification heads**: 
Three different classification heads (Linear, MLP, and attention) were employed, and the differences between the three were compared in subsequent experiments.
![1279ad569f1822b18bb45476177f031](https://github.com/kyotovision/he_sicheng/assets/56251275/002ea16b-ec36-479d-9e0a-abe9f4eee3cb)

### B. Results:
![40d233ac3cd9a0270c825051a2592e1](https://github.com/kyotovision/he_sicheng/assets/56251275/7668cf1b-4c82-454b-a28f-54a3accd1f3b)

- The current classification accuracy with the diffusion model is not as good as the result of classification with the resnet18 network: 0.05053.
- The classification accuracy is increasing as the diffusion model is trained. When the generative model works better, the classification model works better.

## 3. Improvements that need to be made:

To increase the accuracy of this classification model, I think the following points need to be considered:

- Suitable image enhancement and image preprocessing.
When generating standard fonts, it is necessary to increase the robustness for different conditioned images. For example, handwritten kanji are rotated at different angles.

- Classifier-based diffusion model.
In related classification studies based on diffusion model, most of them use classifier guidance. (i.e., a classifier A is pre-trained, and when generating images of a specified category c, the images generated at each step are fed into A to get the classification result c_, which helps to generate more accurate images by penalizing the inconsistency between c and c_.)
However, this approach leverages well-trained discriminative models, and it is not clear to me whether it can be used in my research.

- Question: Should I perform some special processing for Kanji, e.g. with the help of strokes, etc.?
Since currently I have some situation when training diffusion model: for some Kanji, the first few epochs will generate better results than the latest epoch. I think it would be helpful to utilize better prior knowledge related to Kanji to improve this.

**Comment by YunyaoShen** [comment]
你好 我想问下一下 [Diffusion Models Beat GANs on Image Classification](https://arxiv.org/pdf/2307.08702v1.pdf) 这篇文章的源代码在哪里可以找到
